newrelic/node-newrelic
# refactor!: update instrumentation registration to work for CJS deps in ESM without use of a loader

## Description
#1646 revealed that with the introduction of version 10 and the ability to add multiple instrumentation for a given package, that for ESM applications there were edge cases where we could double instrument the package, specifically if the ESM application and a dependency both used the same package. Additionally, @bizob2828's [extensive research](https://github.com/newrelic/node-newrelic/issues/1720#issuecomment-1654614418) into the changes Node 20 made with regards to loader hooks and worker threads culminated in the decision that we won't be continuing down the loader path for Node 20, and will be dropping support for registering custom ESM instrumentation.

This change replaces the major use case of the loader, which is registering CJS package instrumentation in an ESM application. All of the packages we currently implement are CJS packages (express/pg/koa/etc.). As such, and due to our wrapping `Module` internals, we're actually able to go through common paths regardless of whether or not an app is CJS or ESM.

With this change, we detect if we're in an ESM import (which we're able to do because the module name is the same fully resolved filepath in ESM, which was why we originally introduced the loader in the first place), and instead of using the fully resolve filepath as our instrumentation registration key, we use `module-details-from-path` to reverse lookup the package's "short name" from the file path (IE: given `/this/is/a/fully/resolved/file/path/to/node_modules/express/index.js`, we'll get `express` as the "short name"). This allows us to make sure we don't double instrument packages.

Additionally, I've removed the ESM loader from the code base, as in Node 20 it will not work at all. A sibling PR exists for [test-utils](https://github.com/newrelic/node-test-utilities/pull/178) that gates using the loader in versioned tests through an env var. This PR's CI won't pass until the test utils pr is merged and published, but can be vetted locally through `npm link`.

## How to Test
* Manually tested using `npm link` and the following repro apps: https://github.com/wiked03/newrelic-issue and https://github.com/jmartin4563/newrelic-esm-issue
* Added to versioned test harness since that's where the test utility for loaders lives, copy folder to main and run, observe failure, run with `npm link` on https://github.com/newrelic/node-test-utilities/pull/178, observe tests passing

## Related Issues
Closes #1646
